gapic-common/lib/gapic/lru_hash.rb

```diff
@@ -0,0 +1,101 @@
+# Copyright 2023 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+module Gapic
+  ##
+  # Linked list based hash maintaining the order of
+  # access/creation of the keys.
+  #
+  class LruHash
+    def initialize size = 1
+      raise ArgumentError, "The size of LRU hash can't be < 1" unless size > 1
+      @start = nil
+      @end = nil
+      @size = size
+      @cache = {}
+    end
+
+    def get key
+      return nil unless @cache.key? key
+      node = @cache[key]
+      move_to_top node
+      node.value
+    end
+
+    def put key, value
+      if @cache.key? key
+        node = @cache[key]
+        node.value = value
+        move_to_top node
+      else
+        remove_tail if @cache.size >= @size
+        new_node = Node.new key, value
+        insert_at_top new_node
+        @cache[key] = new_node
+      end
+    end
+
+    private
+
+    def move_to_top node
+      return if node == @start
```

Comment:
`node.equal? @start` better expresses this intent. `==` tests value equality and could be true for two distinct objects with the same semantic value. `equal?` tests object identity which is what we're looking for here. Same a couple lines down.

Reply:
Thanks for the explanation. Updated the code to use `.equals?`